element_facemap/facial_behavior_estimation.py

```diff
@@ -0,0 +1,361 @@
+                nframes += int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
+
+        self.insert1({
+            **key,
+            'px_height': px_height,
+            'px_width': px_width,
+            'nframes': nframes,
+            'fps': fps,  # Assuming that this keyword is set correctly but it's usually not.
+            'duration': nframes / fps,                     # TODO Something to think about.
+        })
+    
+
+@schema
+class FacemapParamSet(dj.Manual):
+    definition = """
+    # Parameters used to run the Facemap on each Recording
+    -> VideoRecording
+    paramset_idx        :  smallint
+    ---
+    params              : longblob
+    param_set_hash      : uuid          # unique index
+    do_mot_svd=1        : bool
+    do_mov_svd=0        : bool
+    paramset_desc=''    : varchar(128)
+    """
+
+
+@schema
+class FacemapTask(dj.Manual):
+    definition = """
+    # Manual table for defining a processing task ready to be run
+    -> FacemapParamSet
+    ---
+    processing_output_dir   : varchar(255)              # output directory that will keep the facemap results
+    task_mode='load'        : enum('load', 'trigger')   # 'load': load computed analysis results, 'trigger': trigger computation
+    """
+
+    @classmethod
+    def infer_output_dir(cls, video_key,  relative=False, mkdir=False):
+        video_file = (FacemapParamSet * VideoRecording.File & video_key).fetch('file_path')[0]  # Take 1 video file
+        video_dir = find_full_path(get_facemap_root_data_dir(), video_file)  # find video file's full path
+        root_dir = find_root_directory(get_facemap_root_data_dir(), video_dir)  # find the video file's root directory
+
+        paramset_key = FacemapParamSet.fetch1()
+        processed_dir = Path(get_facemap_processed_data_dir())
+        output_dir = processed_dir / video_dir.relative_to(root_dir) / f'facemap_{paramset_key["paramset_idx"]}'
+
+        if mkdir:
+            output_dir.mkdir(parents=True, exist_ok=True)
+        
+        return output_dir.relative_to(processed_dir) if relative else output_dir
+
+    @classmethod
+    def auto_generate_entries(cls, video_key, task_mode):
+        ### TODO: fill this part.
+        ### Should we do this?
+
+        output_dir = cls.infer_output_dir(video_key, relative=False, mkdir=True)
+        pass 
+    
+
+
+@schema
+class FacemapProcessing(dj.Computed):
+    definition = """
+    # Processing Procedure
+    -> FacemapTask
+    ---
+    processing_time     : datetime  # time of generation of the facemap results
+    package_version=''  : varchar(16)
+    """
+    
+    # Processing only the VideoRecordings that have their Info inserted.
+    @property
+    def key_source(self):
+        return FacemapTask & RecordingInfo.File
+
+    def make(self, key):
+        task_mode = (FacemapTask & key).fetch1('task_mode')
+
+        output_dir = (FacemapTask & key).fetch1('processing_output_dir')
+
+        if not output_dir:
```

Comment:
I'm not sure this would run if `processing_output_dir` is not nullable in the `FacemapTask` table definition

Reply:
empty string is falsy in python - `bool('')` evaluates to False

Reply:
Yup, not relevant anymore.